src/plugins/lua/known_senders.lua

```diff
@@ -212,6 +213,67 @@ local function known_senders_callback(task)
   end
 end
 
+local function verify_local_replies_set(task)
+  local sender = task:get_reply_sender()
+  if not sender then
+    lua_util.debugm(N, task, 'Could not get sender')
+    return nil
+  end
+
+  local sender_string = lua_util.maybe_obfuscate_string(tostring(sender), settings, settings.sender_prefix)
+  local sender_key = make_key_replies(sender_string:lower(), 8)
+
+  local function redis_zscore_local_cb(err, data)
+    if err ~= nil then
+      rspamd_logger.errx(task, 'Couldn\'t find sender %s local replies set with error: %s', sender_key, err)
+      return
+    end
+    rspamd_logger.infox(task, 'Recipient: %s', data[2])
+    if next(data[2]) == nil then
+      rspamd_logger.infox(task, 'Recipients was not verified')
+      return
+    end
+  end
+
+  local function redis_zscore_local_last_cb(err, data)
+    if err ~= nil then
+      rspamd_logger.errx(task, 'Couldn\'t find sender %s local replies set with error: %s', sender_key, err)
+      return
+    end
+    rspamd_logger.infox(task, 'Recipient: %s', data[2])
+    if next(data[2]) == nil then
+      rspamd_logger.infox(task, 'Recipients was not verified')
+      return
+    end
+    rspamd_logger.infox(task, 'Recipients was verified')
+    task:insert_result(settings.symbol_check_mail_local, 1.0, sender_key)
+  end
+
+  local recipients = task:get_recipients('mime')
+
+  for i, rcpt in pairs(recipients) do
+    if i ~= #recipients then
+      lua_redis.redis_make_request(task,
+              redis_params,
+              sender_key,
+              false,
+              redis_zscore_local_cb,
+              'ZSCAN',
```

Comment:
I really don't understand why do we need `ZSCAN` here. It works as O(N) and I get no glue why can't we just use `ZSCORE <sender_key> <rcpt.addr>`.

Reply:
Erm, we don't even need this loop as well - there is `ZMSCORE` command in Redis. Probably, it will not work on ancient Redis though. We also need to limit number of recipients to be checked (10-15 at max). Redis calls are costly, so we cannot overuse them. If `ZMSCORE` is not good from the terms of portability, we can emulate it via a simple Redis script that calls for `ZSCORE` several times. I would anyway avoid extra calls to Redis and, for sure, O(N) commands such as `ZSCAN ... PATTERN` unless absolutely needed.